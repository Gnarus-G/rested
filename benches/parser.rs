use criterion::{criterion_group, criterion_main, Criterion};
use rested::parser;

const SOURCE: &str = r#"
set BASE_URL env("base")

[env("token")]

let string = "death"

post /todos {
   header "Authorization" env("token")
   body json({
    userId: 1,
    id: 999,
    "title": `delectus aut ${string}`,
    "completed": env("base")
  })
}
"#;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse ast", |b| {
        b.iter(|| {
            let mut parser = parser::Parser::new(SOURCE);
            let _ = parser.parse();
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
