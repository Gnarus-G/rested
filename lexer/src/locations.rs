use serde::Serialize;

use crate::Token;

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub struct Location {
    pub line: usize,
    pub col: usize,
}

impl Location {
    pub fn to_end_of(self, span: Span) -> Span {
        Span {
            start: self,
            end: span.end,
        }
    }

    pub fn is_before(self, location: Location) -> bool {
        if self.line == location.line {
            return self.col <= location.col;
        }
        return self.line < location.line;
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub struct Span {
    pub start: Location,
    pub end: Location,
}

impl Span {
    pub fn new(start: Location, end: Location) -> Self {
        Self { start, end }
    }
    pub fn extend_to(self, end: Location) -> Span {
        Span {
            start: self.start,
            end,
        }
    }

    pub fn to_end_of(self, other_span: Span) -> Span {
        Span {
            start: self.start,
            end: other_span.end,
        }
    }

    pub fn width(&self) -> usize {
        let left = self.start.col;
        let right = self.end.col;

        if right > left {
            return right - left;
        }

        return left - right;
    }
}

impl<'source> Into<Span> for Token<'source> {
    fn into(self) -> Span {
        self.span()
    }
}

impl<'source> Into<Span> for &Token<'source> {
    fn into(self) -> Span {
        self.span()
    }
}

pub trait GetSpan {
    fn span(&self) -> Span;
}

impl<'source> GetSpan for Token<'source> {
    fn span(&self) -> Span {
        Span {
            start: self.start.clone(),
            end: Location {
                line: self.start.line,
                col: self.start.col + self.text.len(),
            },
        }
    }
}

pub trait GetSpanOption {
    fn get_span(&self) -> Option<Span>;
}

impl<T: GetSpan> GetSpanOption for Vec<T> {
    fn get_span(&self) -> Option<Span> {
        match (self.first(), self.last()) {
            (None, None) => None,
            (None, Some(e)) => Some(e.span()),
            (Some(e), None) => Some(e.span()),
            (Some(f), Some(l)) => Some(f.span().to_end_of(l.span())),
        }
    }
}
