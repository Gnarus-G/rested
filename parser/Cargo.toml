[package]
name = "parser"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
colored = "2.0.0"
serde = { version = "1.0.159", features = ["derive"] }
lexer = { path = "../lexer" }
error_meta = { path = "../error_meta" }

[dev-dependencies]
insta = { version = "1.29.0", features = ["ron"] }
