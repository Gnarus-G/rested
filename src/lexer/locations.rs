use serde::Serialize;

use super::Token;

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Default)]
pub struct Position {
    pub value: usize,
    pub line: usize,
    pub col: usize,
}

impl From<Position> for Location {
    fn from(value: Position) -> Self {
        Self {
            line: value.line,
            col: value.col,
        }
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Default)]
pub struct Location {
    pub line: usize,
    pub col: usize,
}

impl Position {
    pub fn to_end_of(self, span: Span) -> Span {
        Span {
            start: self,
            end: span.end,
        }
    }
}

impl Location {
    pub fn is_before(self, location: Location) -> bool {
        if self.line == location.line {
            return self.col <= location.col;
        }
        self.line < location.line
    }
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub struct Span {
    pub start: Position,
    pub end: Position,
}

impl Span {
    pub fn new(start: Position, end: Position) -> Self {
        Self { start, end }
    }

    pub fn to_end_of(self, other_span: Span) -> Span {
        Span {
            start: self.start,
            end: other_span.end,
        }
    }

    pub fn width(&self) -> usize {
        let left = self.start.col;
        let right = self.end.col;

        if right > left {
            return right - left;
        }

        left - right
    }
}

pub trait GetSpan {
    fn span(&self) -> Span;
}

impl<'source> GetSpan for Token<'source> {
    fn span(&self) -> Span {
        Span {
            start: self.start,
            end: self.end_position(),
        }
    }
}
