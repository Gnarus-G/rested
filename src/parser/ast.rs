use std::fmt::Display;

use serde::Serialize;

use crate::{
    error_meta::ContextualError,
    lexer::{
        locations::{GetSpan, Position, Span},
        Token,
    },
    utils,
};

use self::result::ParsedNode;

use super::error::ParseError;

type Error<'source> = ContextualError<ParseError<'source>>;

#[derive(Debug, PartialEq, Serialize)]
pub struct Program<'i> {
    pub source: &'i str,
    pub items: utils::Array<Item<'i>>,
}

impl<'i> Program<'i> {
    pub fn new(source: &'i str, items: Vec<Item<'i>>) -> Self {
        Self {
            source,
            items: items.into(),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub struct Literal<'i> {
    pub value: &'i str,
    pub span: Span,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct StringLiteral<'source> {
    pub raw: &'source str,
    pub value: &'source str,
    pub span: Span,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct Block<'source> {
    pub statements: Vec<Statement<'source>>,
    pub span: Span,
}

#[derive(Debug, PartialEq, Serialize)]
pub enum Item<'source> {
    Set {
        identifier: ParsedNode<'source, Token<'source>>,
        value: Expression<'source>,
    },
    Let {
        identifier: ParsedNode<'source, Token<'source>>,
        value: Expression<'source>,
    },
    LineComment(Literal<'source>),
    Request {
        method: RequestMethod,
        endpoint: Endpoint<'source>,
        block: Option<Block<'source>>,
        span: Span,
    },
    Expr(Expression<'source>),
    Attribute {
        location: Position,
        identifier: ParsedNode<'source, Token<'source>>,
        parameters: Option<Arguments<'source>>,
    },
    Error(Box<Error<'source>>),
}

#[derive(Debug, PartialEq, Clone, Copy, Serialize)]
pub enum RequestMethod {
    GET,
    POST,
    DELETE,
    PATCH,
    PUT,
}

impl Display for RequestMethod {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub enum Statement<'i> {
    Header {
        name: ParsedNode<'i, StringLiteral<'i>>,
        value: Expression<'i>,
    },
    Body {
        value: Expression<'i>,
        start: Position,
    },
    LineComment(Literal<'i>),
    Error(Box<Error<'i>>),
}

#[derive(Debug, PartialEq, Serialize)]
pub struct Arguments<'source> {
    pub span: Span,
    pub parameters: Vec<Expression<'source>>,
}

impl<'source> Arguments<'source> {
    pub fn iter(&self) -> impl Iterator<Item = &Expression<'source>> {
        self.parameters.iter()
    }
}

pub type Spanned<T> = (Span, T);

#[derive(Debug, PartialEq, Serialize)]
pub enum Expression<'source> {
    Identifier(ParsedNode<'source, Token<'source>>),
    String(StringLiteral<'source>),
    Bool(Literal<'source>),
    Number(Literal<'source>),
    Call {
        identifier: ParsedNode<'source, Token<'source>>,
        arguments: Arguments<'source>,
    },
    Array(Spanned<Vec<ArrayElement<'source>>>),
    Object(Spanned<Vec<ObjectEntry<'source>>>),
    Null(Span),
    EmptyArray(Span),
    EmptyObject(Span),
    TemplateSringLiteral {
        span: Span,
        parts: Vec<Expression<'source>>,
    },
    Error(Box<Error<'source>>),
}

#[derive(Debug, PartialEq, Serialize)]
pub struct ArrayElement<'source> {
    pub expr: Expression<'source>,
    // For missing commas actually. I don't know if this is a good idea generally though.
    pub errors: Vec<Error<'source>>,
}

#[derive(Debug, PartialEq, Serialize)]
pub struct ObjectEntry<'source> {
    pub key: ParsedNode<'source, StringLiteral<'source>>,
    pub value: Expression<'source>,
    // For missing commas...
    pub errors: Vec<Error<'source>>,
}

impl<'source> ObjectEntry<'source> {
    pub fn new(
        key: ParsedNode<'source, StringLiteral<'source>>,
        value: Expression<'source>,
    ) -> Self {
        Self {
            key,
            value,
            errors: vec![],
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub enum Endpoint<'i> {
    Url(Literal<'i>),
    Pathname(Literal<'i>),
}

pub mod result {

    use super::*;

    #[derive(Debug, PartialEq, serde::Serialize)]
    pub enum ParsedNode<'i, T: GetSpan> {
        Ok(T),
        Error(Box<Error<'i>>),
    }

    impl<'source, T: GetSpan> ParsedNode<'source, T> {
        pub fn get(&self) -> std::result::Result<&T, Box<Error<'source>>> {
            match self {
                ParsedNode::Ok(node) => Ok(node),
                ParsedNode::Error(error) => Err(error.clone()),
            }
        }
    }
}

mod convert {
    use super::*;

    impl<'i> From<&Token<'i>> for ParsedNode<'i, Token<'i>> {
        fn from(token: &Token<'i>) -> Self {
            Self::Ok(Token {
                kind: token.kind,
                text: token.text,
                start: token.start,
            })
        }
    }
    impl<'i> From<&Token<'i>> for Literal<'i> {
        fn from(token: &Token<'i>) -> Self {
            Self {
                value: token.text,
                span: token.span(),
            }
        }
    }
    impl<'i> From<&Token<'i>> for StringLiteral<'i> {
        fn from(token: &Token<'i>) -> Self {
            let value = match (token.text.chars().next(), token.text.chars().last()) {
                (Some('"'), Some('"')) if token.text.len() > 1 => {
                    &token.text[1..token.text.len() - 1]
                }
                (Some('`'), Some('`')) if token.text.len() > 1 => {
                    &token.text[1..token.text.len() - 1]
                }
                (_, Some('`')) => &token.text[..token.text.len() - 1],
                (Some('`'), _) => &token.text[1..],
                _ => token.text,
            };

            Self {
                raw: token.text,
                value,
                span: token.span(),
            }
        }
    }

    impl<'source, T: GetSpan> From<std::result::Result<T, Box<Error<'source>>>>
        for ParsedNode<'source, T>
    {
        fn from(value: std::result::Result<T, std::boxed::Box<Error<'source>>>) -> Self {
            match value {
                Ok(value) => ParsedNode::Ok(value),
                Err(error) => ParsedNode::Error(error),
            }
        }
    }

    impl<'source> From<Expression<'source>> for ArrayElement<'source> {
        fn from(value: Expression<'source>) -> Self {
            Self {
                expr: value,
                errors: vec![],
            }
        }
    }

    impl<'source> From<Error<'source>> for Expression<'source> {
        fn from(value: Error<'source>) -> Self {
            Self::Error(value.into())
        }
    }

    impl<'source> From<Error<'source>> for Statement<'source> {
        fn from(value: Error<'source>) -> Self {
            Self::Error(value.into())
        }
    }

    impl<'source> From<Box<Error<'source>>> for Statement<'source> {
        fn from(value: Box<Error<'source>>) -> Self {
            Self::Error(value)
        }
    }

    impl<'source> From<Error<'source>> for Item<'source> {
        fn from(value: Error<'source>) -> Self {
            Self::Error(value.into())
        }
    }

    impl<'source> From<std::result::Result<Self, Box<Error<'source>>>> for Expression<'source> {
        fn from(value: std::result::Result<Self, Box<Error<'source>>>) -> Self {
            match value {
                Ok(v) => v,
                Err(error) => Self::Error(error),
            }
        }
    }
}
