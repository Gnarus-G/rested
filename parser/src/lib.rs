pub mod ast;
mod ast_queries;
mod ast_span;
pub mod error;

use std::collections::BTreeMap;

use crate::ast::{Endpoint, Expression, Item, Program, RequestMethod, Statement};
use ast::Block;
use error::ParserErrors;
use error_meta::ContextualError;
use lexer::{
    locations::{GetSpan, Location, Span},
    Lexer, Token, TokenKind,
};

use self::error::{ParseError, ParseErrorConstructor};

use TokenKind::*;

macro_rules! match_or_throw {
    ($expression:expr; $self:ident; $( $( $pattern:ident )|+ $( if $guard: expr )? => $arm:expr $(,)? )+ $( ,$message:literal )? ) => {
        match $expression {
            $(
                $( TokenKind::$pattern )|+ $( if $guard )? => $arm,
            )+
            _ => return Err($self.error().expected_one_of_tokens($self.curr_token(), vec![$( $( $pattern ),+ ),+])$(.with_message($message))?)
        }
    };
}

pub type Result<'source, T> = std::result::Result<T, ContextualError<ParseError<'source>>>;

trait TokenCheck {
    fn is_one_of(&self, kinds: &[TokenKind]) -> bool;
    fn is(&self, kind: TokenKind) -> bool;
}

impl<'source> TokenCheck for Token<'source> {
    fn is_one_of(&self, kinds: &[TokenKind]) -> bool {
        kinds.contains(&self.kind)
    }

    fn is(&self, kind: TokenKind) -> bool {
        self.kind == kind
    }
}

#[derive(Debug)]
pub struct Parser<'i> {
    lexer: Lexer<'i>,
    token: Option<Token<'i>>,
    peeked: Option<Token<'i>>,
}

impl<'i> Parser<'i> {
    pub fn new(code: &'i str) -> Self {
        Self {
            peeked: None,
            lexer: Lexer::new(code),
            token: None,
        }
    }

    fn curr_token(&self) -> &Token<'i> {
        &self
            .token
            .as_ref()
            .expect("self.token should be initialized at the start of parsing")
    }

    fn next_token(&mut self) -> &Token<'i> {
        self.token = match self.peeked.take() {
            Some(t) => Some(t),
            None => Some(self.lexer.next()),
        };
        self.curr_token()
    }

    fn peek_token(&mut self) -> &Token<'i> {
        self.peeked.get_or_insert_with(|| self.lexer.next())
    }

    fn eat_till_next_top_level_peek_token(&mut self) {
        loop {
            let is_top_level_token_ahead = match self.peek_token().kind {
                Get | Post | Put | Patch | Delete | Set | AttributePrefix | Let | End => true,
                _ => false,
            };

            if is_top_level_token_ahead {
                break;
            }

            self.next_token();
        }
    }

    fn span_from(&self, start: Location) -> Span {
        start.to_end_of(self.curr_token().span())
    }

    pub fn parse(&mut self) -> std::result::Result<Program<'i>, ParserErrors<'i>> {
        let mut program = Program::new();
        let mut errors = vec![];

        use lexer::TokenKind::*;

        self.next_token();

        while self.curr_token().kind != End {
            let result = match self.curr_token().kind {
                Get => self.parse_request(RequestMethod::GET),
                Post => self.parse_request(RequestMethod::POST),
                Put => self.parse_request(RequestMethod::PUT),
                Patch => self.parse_request(RequestMethod::PATCH),
                Delete => self.parse_request(RequestMethod::DELETE),
                Linecomment | Shebang => Ok(Item::LineComment(self.curr_token().into())),
                Set => self.parse_set_statement(),
                AttributePrefix => {
                    let item = self.parse_attribute();

                    if let Ok(_) = item {
                        let valid_after_attribute =
                            vec![Get, Post, Put, Patch, Delete, AttributePrefix, Linecomment];

                        if let Err(err) = self.expect_peek_one_of(valid_after_attribute) {
                            errors.push(err.with_message(
                                "after attributes should come requests or more attributes",
                            ));
                            continue;
                        }
                    }

                    item
                }
                Let => self.parse_let_statement(),
                _ => match self.parse_expression() {
                    Ok(exp) => Ok(Item::Expr(exp)),
                    Err(err) => Err(err),
                },
            };

            match result {
                Ok(item) => program.items.push(item),
                Err(error) => {
                    errors.push(error);
                    self.eat_till_next_top_level_peek_token();
                }
            }

            self.next_token();
        }

        if errors.is_empty() {
            return Ok(program);
        }

        return Err(ParserErrors::new(errors, program));
    }

    fn parse_request(&mut self, method: RequestMethod) -> Result<'i, Item<'i>> {
        let start = self.curr_token().start;

        let url = self.next_expecting_one_of(vec![Url, Pathname])?;

        let endpoint = match_or_throw! { url.kind; self;
            Url => Endpoint::Url(url.into()),
            Pathname => Endpoint::Pathname(url.into()),
            "expecting only a url and pathname here"
        };
        let url_span: Span = url.span();

        let block = self.parse_block()?;

        let span_next = if let Some(b) = block.as_ref() {
            b.span
        } else {
            url_span
        };

        Ok(Item::Request {
            span: start.to_end_of(span_next),
            method,
            endpoint,
            block,
        })
    }

    fn parse_set_statement(&mut self) -> Result<'i, Item<'i>> {
        let identifier = self.expect_peek(TokenKind::Ident)?.into();

        self.next_expecting_one_of(vec![
            TokenKind::Ident,
            TokenKind::StringLiteral,
            TokenKind::MultiLineStringLiteral,
        ])?;

        Ok(Item::Set {
            identifier,
            value: self.parse_expression()?,
        })
    }

    fn parse_block(&mut self) -> Result<'i, Option<Block<'i>>> {
        let LBracket = self.peek_token().kind else {
            return Ok(None);
        };

        let span_start = self.next_token().start; // remember LBracket's location
        self.next_token();
        let mut statements = vec![];

        while self.curr_token().kind != RBracket && self.curr_token().kind != End {
            let statement = match_or_throw! { self.curr_token().kind; self;
                Header => self.parse_header()?,
                Body => self.parse_body()?,
                Linecomment | Shebang => Statement::LineComment(self.curr_token().into()),
                "may only declare headers or a body statement here"
            };
            statements.push(statement);
            self.next_token();
        }

        return Ok(Some(Block {
            statements,
            span: Span::new(span_start, self.curr_token().start), // span to RBracket's location
        }));
    }

    fn parse_header(&mut self) -> Result<'i, Statement<'i>> {
        let header_name = self.expect_peek(TokenKind::StringLiteral)?.into();

        self.next_expecting_one_of(vec![
            TokenKind::StringLiteral,
            TokenKind::Ident,
            TokenKind::MultiLineStringLiteral,
        ])?;

        let value = self.parse_expression()?;

        Ok(Statement::Header {
            name: header_name,
            value,
        })
    }

    fn parse_body(&mut self) -> Result<'i, Statement<'i>> {
        let start = self.curr_token().start;

        self.next_token();

        let value = self.parse_expression()?;

        Ok(Statement::Body { value, start })
    }

    fn parse_expression(&mut self) -> Result<'i, Expression<'i>> {
        let kind = self.curr_token().kind;

        let exp = match_or_throw! { kind; self;
            Ident if self.peek_token().kind == LParen => self.parse_call_expression()?,
            Ident => Expression::Identifier(self.curr_token().into()),
            StringLiteral => Expression::String(self.curr_token().into()),
            Boolean => Expression::Bool(self.curr_token().into()),
            Number => Expression::Number(self.curr_token().into()),
            MultiLineStringLiteral => self.parse_multiline_string_literal()?,
            LBracket | LSquare => self.parse_json_like()?,
            Null => Expression::Null(self.curr_token().span()),
        };

        Ok(exp)
    }

    fn parse_json_like(&mut self) -> Result<'i, Expression<'i>> {
        let start_token = self.curr_token();
        let start = start_token.start;

        let object = match start_token.kind {
            LBracket => {
                if self.peek_token().kind == RBracket {
                    self.next_token();
                    return Ok(Expression::EmptyObject(self.span_from(start)));
                }

                let mut fields = BTreeMap::new();

                while self.peek_token().kind != RBracket {
                    if self.peek_token().is(Linecomment) {
                        self.next_token();
                        continue;
                    }

                    let (key, value) = self.parse_object_property()?;
                    fields.insert(key, value);

                    if !self.peek_token().is(RBracket) {
                        self.expect_peek(Comma)?;
                    }
                }

                self.next_token();

                let span = start.to_end_of(self.curr_token().span());

                Expression::Object((span, fields))
            }
            LSquare => {
                if self.peek_token().kind == RSquare {
                    self.next_token();
                    return Ok(Expression::EmptyArray(self.span_from(start)));
                }

                let mut list = vec![];

                while self.peek_token().kind != RSquare {
                    if self.peek_token().is(Linecomment) {
                        self.next_token();
                        continue;
                    }

                    self.next_token();
                    list.push(self.parse_json_like()?);

                    if !self.peek_token().is(RSquare) {
                        self.expect_peek(Comma)?;
                    }
                }

                self.next_token();

                let span = start.to_end_of(self.curr_token().span());

                Expression::Array((span, list))
            }
            _ => self.parse_expression()?,
        };

        Ok(object)
    }

    fn parse_object_property(&mut self) -> Result<'i, (&'i str, Expression<'i>)> {
        let key_token = self.next_token();

        let key = match_or_throw! { key_token.kind; self;
            Ident => key_token.text,
            StringLiteral => ast::StringLiteral::from(key_token).value
        };

        self.expect_peek(TokenKind::Colon)?;
        self.next_token();

        return Ok((key, self.parse_json_like()?));
    }

    fn parse_call_expression(&mut self) -> Result<'i, Expression<'i>> {
        let identifier = self.curr_token().into();
        self.next_token();

        self.next_token();

        let mut arguments = vec![];

        while self.curr_token().kind != TokenKind::RParen {
            let exp = self.parse_expression()?;
            arguments.push(exp);
            self.next_token();
        }

        Ok(Expression::Call {
            identifier,
            arguments,
        })
    }

    fn parse_multiline_string_literal(&mut self) -> Result<'i, Expression<'i>> {
        let mut parts = vec![];
        let start = self.curr_token().start;
        let end;

        loop {
            let kind = self.curr_token().kind;

            match_or_throw! { kind; self;
                MultiLineStringLiteral
                    if self.peek_token().kind == TokenKind::DollarSignLBracket =>
                {
                    let s_literal = Expression::String(self.curr_token().into());

                    parts.push(s_literal);

                    self.next_token();

                    self.next_token();

                    parts.push(self.parse_expression()?);
                }
                End => {
                    end = self.curr_token().end_location();
                    break;
                },
                MultiLineStringLiteral if parts.is_empty() => {
                    return Ok(Expression::String(self.curr_token().into()));
                }
                MultiLineStringLiteral => {
                    end = self.curr_token().end_location();
                    parts.push(Expression::String(self.curr_token().into()));
                    break;
                }
            };

            self.next_token();
        }

        Ok(Expression::TemplateSringLiteral {
            span: Span::new(start, end),
            parts,
        })
    }

    fn next_expecting_one_of(&mut self, expected_kinds: Vec<TokenKind>) -> Result<'i, &Token<'i>> {
        self.expect_peek_one_of(expected_kinds)
            .map(|_| self.next_token())
    }

    fn expect_peek_one_of(&mut self, expected_kinds: Vec<TokenKind>) -> Result<'i, ()> {
        if self.peek_token().is_one_of(&expected_kinds) {
            return Ok(());
        }
        let con = self.error();

        let error = con.expected_one_of_tokens(&self.next_token(), expected_kinds);

        Err(error)
    }

    fn expect_peek(&mut self, expected_kind: TokenKind) -> Result<'i, &Token<'i>> {
        if self.peek_token().is(expected_kind) {
            return Ok(self.next_token());
        }

        let error = self
            .error()
            .expected_token(&self.next_token(), expected_kind);

        Err(error)
    }

    fn error(&self) -> ParseErrorConstructor<'i> {
        ParseErrorConstructor::new(self.lexer.input())
    }

    fn parse_attribute(&mut self) -> Result<'i, Item<'i>> {
        let location = self.curr_token().start;

        let identifier = self.expect_peek(TokenKind::Ident)?.into();

        let mut params = vec![];

        if let TokenKind::LParen = self.peek_token().kind {
            self.next_token();

            self.next_token();
            while self.curr_token().kind != TokenKind::RParen {
                let exp = self.parse_expression()?;

                params.push(exp);

                self.next_token();
            }
        }

        Ok(Item::Attribute {
            location,
            identifier,
            parameters: params,
        })
    }

    fn parse_let_statement(&mut self) -> Result<'i, Item<'i>> {
        let identifier = self.next_token().into();

        self.expect_peek(TokenKind::Assign)?;

        self.next_token();

        Ok(Item::Let {
            identifier,
            value: self.parse_expression()?,
        })
    }
}
