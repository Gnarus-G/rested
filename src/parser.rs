use crate::{
    ast::{Header, Program, Request},
    error::{ParseError, ParseErrorKind},
    lexer::{Lexer, Token, TokenKind},
};

type Result<T> = std::result::Result<T, ParseError>;

#[derive(Debug)]
pub struct Parser<'a> {
    lexer: Lexer<'a>,
    peeked: Option<Token<'a>>,
}

impl<'i> Parser<'i> {
    pub fn new(lexer: Lexer<'i>) -> Self {
        Self {
            lexer,
            peeked: None,
        }
    }

    fn token(&mut self) -> Token<'i> {
        match self.peeked.take() {
            Some(t) => t,
            None => self.lexer.next(),
        }
    }

    fn peek_token(&mut self) -> &Token<'i> {
        self.peeked.get_or_insert_with(|| self.lexer.next())
    }

    fn eat_token(&mut self) {
        self.token();
    }

    pub fn parse(&mut self) -> Result<Program<'i>> {
        let mut program = Program::new();

        use crate::lexer::TokenKind::*;

        let mut token = self.token();

        while token.kind != End {
            let request = match token.kind {
                Get => self.parse_get_request()?,
                tk => todo!("{tk:?}"),
            };
            program.requests.push(request);
            token = self.token();
        }

        Ok(program)
    }

    fn parse_get_request(&mut self) -> Result<Request<'i>> {
        Ok(Request::Get(crate::ast::GetRequestParams {
            url: self.token().text,
            headers: self.parse_get_params()?,
        }))
    }

    fn parse_get_params(&mut self) -> Result<Option<Vec<Header<'i>>>> {
        let mut token = self.token();
        if let crate::lexer::TokenKind::LBracket = token.kind {
            token = self.token();
            let mut headers = vec![];
            while token.kind != TokenKind::RBracket {
                let h = match token.kind {
                    TokenKind::Header => self.parse_header()?,
                    TokenKind::Ident => todo!(),
                    TokenKind::StringLiteral => todo!(),
                    TokenKind::Assign => todo!(),
                    TokenKind::Quote => todo!(),
                    tk => todo!("{tk:?}"),
                };

                headers.push(h);
                token = self.token();
            }
            Ok(Some(headers))
        } else {
            Ok(None)
        }
    }

    fn parse_header(&mut self) -> Result<Header<'i>> {
        let t = self.token();

        let header_name = t.text;

        self.expect(TokenKind::Assign)?;

        self.eat_token();

        self.expect(TokenKind::Quote)?;

        self.eat_token();

        self.expect(TokenKind::StringLiteral)?;

        let header_value = self.token().text;

        self.expect(TokenKind::Quote)?;

        self.eat_token();

        Ok(Header {
            name: header_name,
            value: header_value,
        })
    }

    fn expect(&mut self, kind: TokenKind) -> Result<()> {
        let ahead = self.peek_token();

        if ahead.kind == kind {
            return Ok(());
        }

        Err(ParseError::new(
            ParseErrorKind::Unexpected {
                expected: kind,
                found: ahead.kind,
            },
            ahead.location,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::ast::{GetRequestParams, Header, Program, Request};

    macro_rules! assert_program {
        ($input:literal, $program:expr) => {
            let lexer = Lexer::new($input);
            let mut parser = Parser::new(lexer);
            assert_eq!(parser.parse().unwrap(), $program);
        };
    }

    #[test]
    fn parse_get_url() {
        assert_program!(
            "get http://localhost",
            Program {
                requests: vec![Request::Get(GetRequestParams {
                    url: "http://localhost",
                    headers: None
                })]
            }
        );
    }

    #[test]
    fn parse_get_with_headers() {
        assert_program!(
            "get http://localhost { header Authorization = \"Bearer token\" }",
            Program {
                requests: vec![Request::Get(GetRequestParams {
                    url: "http://localhost",
                    headers: Some(vec![Header {
                        name: "Authorization",
                        value: "Bearer token"
                    }])
                })]
            }
        );
    }
}
