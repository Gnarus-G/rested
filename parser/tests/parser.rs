use insta::assert_ron_snapshot;

use parser::Parser;

macro_rules! assert_ast {
    ($input:literal) => {
        let mut parser = Parser::new($input);
        let ast = parser.parse().unwrap();

        insta::with_settings!({
             description => $input
        }, {
            assert_ron_snapshot!(ast);
        })
    };
}

#[test]
fn it_works() {
    assert_ast!(
        r#"
set BASE_URL "httas..."
post http://lasdf.. {}
// asdfasdf

let output_file = "output/file.json"
let token = env("auth-token")

@log(output_file)
post /asd {
  // asdfasd
  header "Authorization" token
  body `{
      "neet": "${env("var")}",
      "nerd": "${output_file}",
  }`
}

@skip
get /api {}

put /api {}
patch /api {}
delete /api {}
"#
    );
}

use lexer::locations::Location;

pub fn at(line: usize, col: usize) -> Location {
    Location { line, col }
}

#[test]
fn parse_get_urls() {
    assert_ast!(
        r#"get http://localhost:8080
        get http://localhost:8080 {}"#
    );
}

#[test]
fn parse_post_url() {
    assert_ast!("post http://localhost");

    assert_ast!("post /api/v2");
}

#[test]
fn parse_attributes() {
    assert_ast!(r#"@log("path/to/file") get /api"#);
}

#[test]
fn parse_get_with_headers() {
    assert_ast!(
        r#"
        get http://localhost {
        header "Authorization" "Bearer token"
        header "random" "tokener Bear"
        }"#
    );
}

#[test]
fn parse_post_with_headers() {
    assert_ast!(
        r#"
        post http://localhost {
        header "Authorization" "Bearer token"
        header "random" "tokener Bear"
        }"#
    );
}
#[test]
fn parse_post_with_headers_and_body() {
    assert_ast!(
        r#"
        post http://localhost {
        header "Authorization" "Bearer token"
        body "{neet: 1337}"
        }"#
    );
}

#[test]
fn parse_post_with_headers_and_body_as_json_string() {
    assert_ast!(
        r#"
        post http://localhost {
        header "Authorization" "Bearer token"
        body `
        {"neet": 1337}
        `
        }"#
    );
}

#[test]
fn parse_env_call_expression() {
    assert_ast!(r#"post http://localhost { header "name" env("auth") body env("data") }"#);
}

#[test]
fn parse_global_constant_setting() {
    assert_ast!("set BASE_URL \"stuff\"");
}

#[test]
fn parse_template_string_literal() {
    assert_ast!(
        r#"
        post /api {
            body `{"neet": ${env("love")}, 2: ${"two"}}`
        }"#
    );
}
