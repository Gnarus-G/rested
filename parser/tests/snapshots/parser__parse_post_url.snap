---
source: parser/tests/parser.rs
description: "post http://localhost"
expression: ast
---
Program(
  items: [
    Request(
      method: POST,
      endpoint: Url(Literal(
        value: "http://localhost",
        span: Span(
          start: Location(
            line: 0,
            col: 5,
          ),
          end: Location(
            line: 0,
            col: 21,
          ),
        ),
      )),
      params: [],
      span: Span(
        start: Location(
          line: 0,
          col: 0,
        ),
        end: Location(
          line: 0,
          col: 21,
        ),
      ),
    ),
  ],
)
